hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsSelectorCommon.java

```diff
@@ -316,6 +343,197 @@ private static Dataset<Row> coalesceOrRepartition(Dataset dataset, int numPartit
+
+    StructField[] modifiedFields = getRecordFields(avroSchema).stream()
+        .flatMap(avroField -> generateRowFieldsWithAliases(avroField, rowFieldsMap.get(avroField.name())).stream())
+        .toArray(StructField[]::new);
+
+    return new StructType(modifiedFields);
+  }
+
+  private static List<Schema.Field> getRecordFields(Schema schema) {
+    if (schema.getType() == Schema.Type.RECORD) {
+      return schema.getFields();
+    }
+
+    if (schema.getType() == Schema.Type.UNION) {
+      return schema.getTypes().stream()
+          .filter(subSchema -> subSchema.getType() == Schema.Type.RECORD)
+          .findFirst()
+          .map(Schema::getFields)
+          .orElse(Collections.emptyList());
+    }
+
+    return Collections.emptyList();
+  }
+
+  /**
+   * Generates a list of StructFields with aliases applied based on the provided Avro field schema.
+   * <p>
+   * This method processes a given Avro field and its corresponding Spark SQL StructField, handling
+   * nested records and aliases. If the Avro field contains nested records, the method recursively
+   * updates the schema for these records and applies any aliases defined in the Avro schema.
+   * If the Avro field has aliases, they are added as new fields with nullable set to true and
+   * appropriate metadata in the returned list. If no aliases or nesting are present, the original
+   * StructField is returned unchanged.
+   *
+   * @param avroField The Avro field schema to process.
+   * @param rowField  The corresponding Spark SQL StructField to map the Avro field to.
+   * @return A list of StructFields with aliases applied as per the Avro schema.
+   */
+  private static List<StructField> generateRowFieldsWithAliases(Schema.Field avroField, StructField rowField) {
+    List<StructField> fieldList = new ArrayList<>();
+
+    // Handle nested records
+    if (isNestedRecord(avroField.schema())) {
+      StructType updatedSchema = addAliasesToRowSchema(avroField.schema(), (StructType) rowField.dataType());
+
+      if (schemaModifiedOrHasAliases(avroField, updatedSchema, rowField)) {
+        // Add the original field with the updated schema and add aliases if present
+        addFieldWithAliases(fieldList, avroField.name(), updatedSchema, rowField.metadata(), avroField.aliases());
+      } else {
+        fieldList.add(rowField);
+      }
+    } else if (!avroField.aliases().isEmpty()) {
+      // If the field has aliases, add them to the schema
+      addFieldWithAliases(fieldList, avroField.name(), rowField.dataType(), rowField.metadata(), avroField.aliases());
+    } else {
+      // No aliases or nesting, return the original field
+      fieldList.add(rowField);
+    }
+    return fieldList;
+  }
+
+  private static void addFieldWithAliases(List<StructField> fieldList, String fieldName, DataType dataType, Metadata metadata, Set<String> aliases) {
+    fieldList.add(new StructField(fieldName, dataType, true, metadata));
```

Comment:
Should we avoid hard-coded nullability and follow the nullability from source schema?

Reply:
If an avro field has an alias set and it can happen that some rows have value with actual field name and some with alias name. so field can be null in such scenarios. Once the dataset is read after applying coalescing, we do apply the source schema on the final data to make sure we're not modifying the original source schema.

`dataset = spark.createDataFrame(coalesceAliasFields(dataset, sourceSchema).rdd(), rowSchema);`

For e.g
lastName | SurName
James.     |  null
null           | James